{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Counting DNA Nucleotides\n",
    "\n",
    "Given: A DNA string s of length at most 1000 nt.\n",
    "\n",
    "Return: Four integers (separated by spaces) counting the respective number of times that the symbols 'A', 'C', 'G', and 'T' occur in s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AATTAACAGCACCCCTAATTTTTGCCAACTAGCGACAAAAAGCGTAGAGCTGCATCAAATCACAGTCCGCTCATAACTGCGTGGTCTTCGACCGGATGATAGTAGCGTACTCCTACAGGCCTGCGGACCAGGTGGAGCCGATGATTCCGGAACTCGCGGCCAGGTGCATCACGATGGGCCTAATAACCTGGCATGGAAAATTGGGTTCAGATTTATTTAAAGGCCCCTCTGATAGAGTTTCGAACAACTGTTTACAGTTACTTTGCTCACGGCATTTGACTCTAGCCGCCCATCGGCACCACTTTCTAGCATCTAACATAAGACACCTCCGCTATTCTCCGCCACTCGGCATATCGATAGTCCCAAACGGCACCGAAAGTTAATTACGGTTAGCTGTGTGTTTGATGTGTACATCTGGACAATGATATCGGCAACTAACATTGAGCTAGATAATTGCACCCAGGATCGGTGCGGATCAAGTTCAATGGTCCGGACCCAAGAGGGGGTTAAGGTAGAGCCAGAAGTTAACTATCGCTCCATTGCAAAAAGACCTAAATCGGGGGTCCGTCCTCCCTTACAGACAGTCCGAGCCGCATAAGCTCTGTTAGCCCTTATAGCCAACTTCCGGGGTACACGGACCAGGCTCACGCTAAATGTCCGCCAAGTATTTCTTGCCCACCCTTTGCACTTAATCGAAGGACTTGTAGTTGAAATAACTAGACGTTACGGCAGTACGGATGAAATTGTAAATTACCACGATCGGTTCCGCTACTGATAAAAAGACTCACACGCATGATCCGTGCTTCCAAATGATGGCTGGTAACCTGTTAGCGCTCTACCATCGCACGTAAGGTGCCAGG\n",
      "A 232\n",
      "T 208\n",
      "C 224\n",
      "G 196\n",
      "232 224 196 208\n"
     ]
    }
   ],
   "source": [
    "s = input()\n",
    "symb = {}\n",
    "for el in s:\n",
    "    if el not in symb:\n",
    "        symb[el] = 1\n",
    "    else:\n",
    "        symb[el] += 1\n",
    "for el in symb:\n",
    "    print(el, symb[el])\n",
    "print( symb['A'], symb['C'], symb['G'], symb['T'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transcribing DNA into RNA\n",
    "\n",
    "Given: A DNA string t having length at most 1000 nt.\n",
    "\n",
    "Return: The transcribed RNA string of t."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACCATCCGTAATGATGCGAGTCCCTCTTTGACGTCCTCAAGCCAGTTGGCAGATCGTTAGGCCGTACACACCGAAATATACCCCACGTTATAGTCGATGTCTAATAGCCATCGGCGGCTGTGTTGGACTTGTTATGGTGAAGGGCCAGAAGTGCTAGCCTAACCCCGACTTGTGAGGAACTTGCCTATCAAACGACGACCACGCGCGGGATTCTCCGCGGTGCGATATTGCGGGTGGCAACTGACGACAGGTATGAGCGAACCTTGGGGTAATGCAGGTTTACCAGTTCGTGACCTACTGTACATGCTTGTAACTTACATGGAACGTCAGCTAAGTACTGGATCCGTTTTCTAAGTTTGGATCGAATAGCAACGGTTTCGTGGCCTGGCGCGCCGTACTAAGAACATTGAAGTGTGGTCCTGCCGTCCAACAGGTTCGCCCGTCATAACGGCGCTGGGGAATCAGCGGCGTCCCTCTTATAGGGGGCGTGCCATACCTGACCCTCTGCATTACGTCTCGAATCGTCTTCGACACGCGGGAGGGCTGGAGAGAGTGGTCTTCATTCACGGATAGAGTGCCGATATCTTGTAAATGCTTGACTAGGAACCGAGACATCATTAGTTAACGAATCTACCGTCAGAGGCAAACTCGTATTCCGTATCCGCATTATCGCATCAAAGTACTGTCCAGGCGTAGATACAAATCTTGTGCCTTTGAACTTATAGCCCTGTTTCTATGGCCTTCACCGGGAAATCCGATTGTGATAAGAGCCGTCGATGTGCTCTACTAAGCCACAATGGATATCTTATAGTTCTAGTCAATTGGAGAACTACACCGTTACGTCCATTGTGCACTACAGCATCGCGAGTCCTTTTAATGCTTTTCGACTACGCGGACCATGAACCCTC\n",
      "CACCAUCCGUAAUGAUGCGAGUCCCUCUUUGACGUCCUCAAGCCAGUUGGCAGAUCGUUAGGCCGUACACACCGAAAUAUACCCCACGUUAUAGUCGAUGUCUAAUAGCCAUCGGCGGCUGUGUUGGACUUGUUAUGGUGAAGGGCCAGAAGUGCUAGCCUAACCCCGACUUGUGAGGAACUUGCCUAUCAAACGACGACCACGCGCGGGAUUCUCCGCGGUGCGAUAUUGCGGGUGGCAACUGACGACAGGUAUGAGCGAACCUUGGGGUAAUGCAGGUUUACCAGUUCGUGACCUACUGUACAUGCUUGUAACUUACAUGGAACGUCAGCUAAGUACUGGAUCCGUUUUCUAAGUUUGGAUCGAAUAGCAACGGUUUCGUGGCCUGGCGCGCCGUACUAAGAACAUUGAAGUGUGGUCCUGCCGUCCAACAGGUUCGCCCGUCAUAACGGCGCUGGGGAAUCAGCGGCGUCCCUCUUAUAGGGGGCGUGCCAUACCUGACCCUCUGCAUUACGUCUCGAAUCGUCUUCGACACGCGGGAGGGCUGGAGAGAGUGGUCUUCAUUCACGGAUAGAGUGCCGAUAUCUUGUAAAUGCUUGACUAGGAACCGAGACAUCAUUAGUUAACGAAUCUACCGUCAGAGGCAAACUCGUAUUCCGUAUCCGCAUUAUCGCAUCAAAGUACUGUCCAGGCGUAGAUACAAAUCUUGUGCCUUUGAACUUAUAGCCCUGUUUCUAUGGCCUUCACCGGGAAAUCCGAUUGUGAUAAGAGCCGUCGAUGUGCUCUACUAAGCCACAAUGGAUAUCUUAUAGUUCUAGUCAAUUGGAGAACUACACCGUUACGUCCAUUGUGCACUACAGCAUCGCGAGUCCUUUUAAUGCUUUUCGACUACGCGGACCAUGAACCCUC\n"
     ]
    }
   ],
   "source": [
    "t = input()\n",
    "#DNA_to_RNA = {'T':'U'}\n",
    "u = t.replace('T', 'U')\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Complementing a Strand of DNA\n",
    "\n",
    "Given: A DNA string s of length at most 1000 bp.\n",
    "\n",
    "Return: The reverse complement sc of s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTTCGCCCAAGGCGCTAGCATATCTTGAACATTCTTTCCCGCCGACGGTGTGTATTACAAGATACCGGCCGGAAGGCTAAAGCTGGCCTCACACCTCTGGCCACAGGCTAAAAGTAGCTCTCGTCTACACTCTTAACCCAACGGGTATCTGAGGGAAACAGCCTTTGTCTGTGGCTGTACGGGCTGCAAAGCCTAGTGGTTAGTCAATTGGATAAGGGCAACTCACGCCTGGAGTCTGACACGATAAGCCGACGCTGACCCGGCCGAATTGTCAGCTCTAAGGGAGGATCAGTCCAATAAGGCGCACGCACCTTGCTGTGGGTCGACCCAAGGAATGGACGACCAACCGTTTACTAGATTCATTAGCTCCTGCAGCCGGAGGCACCTCGCAGAACCCTTGCTACACATCAAAAGACTACTTCAAGATCCACTCACATCCCCCAGAACATATGTGCTTACGGGTGCGAAAACTCAGAGTCATAGCCAGCGAGGTACCAGTAGAGGCAGATGTCGATTTGGACACTCTTAAGAGGAGCGGACTATGTTCCCAGCTTTTGGCGGTATGCTCGCCCGTGTCAACTTAACAATAAATGTTAGCCTGGCTTTCGTACCTAGAGCTACTGCGCCCGCCCGGCACCTGTTCGGTACATAGGTGTGATGAGCCTGGATACCCTGGCATTATAAGTGATGCTGTCCCCCACTGTAGCCTCGTCACGAGCGACCAATTCCCGCTCAGGTCCTTGAGTACAATGCGGCTCGATAGCCTGGGATCATGCGTTGATATGAAGCCAGCTAAGGGTAGTAATCGCCGAGTGCTAAAGTACGACTCCCCTTCAGTCTACGTTCCACTATACGGGGTCTACCTGAGCGGATTCCTCACGGGTCCATGCTAGGTTCATTATAGTACCGTTCACTCT\n",
      "AGAGTGAACGGTACTATAATGAACCTAGCATGGACCCGTGAGGAATCCGCTCAGGTAGACCCCGTATAGTGGAACGTAGACTGAAGGGGAGTCGTACTTTAGCACTCGGCGATTACTACCCTTAGCTGGCTTCATATCAACGCATGATCCCAGGCTATCGAGCCGCATTGTACTCAAGGACCTGAGCGGGAATTGGTCGCTCGTGACGAGGCTACAGTGGGGGACAGCATCACTTATAATGCCAGGGTATCCAGGCTCATCACACCTATGTACCGAACAGGTGCCGGGCGGGCGCAGTAGCTCTAGGTACGAAAGCCAGGCTAACATTTATTGTTAAGTTGACACGGGCGAGCATACCGCCAAAAGCTGGGAACATAGTCCGCTCCTCTTAAGAGTGTCCAAATCGACATCTGCCTCTACTGGTACCTCGCTGGCTATGACTCTGAGTTTTCGCACCCGTAAGCACATATGTTCTGGGGGATGTGAGTGGATCTTGAAGTAGTCTTTTGATGTGTAGCAAGGGTTCTGCGAGGTGCCTCCGGCTGCAGGAGCTAATGAATCTAGTAAACGGTTGGTCGTCCATTCCTTGGGTCGACCCACAGCAAGGTGCGTGCGCCTTATTGGACTGATCCTCCCTTAGAGCTGACAATTCGGCCGGGTCAGCGTCGGCTTATCGTGTCAGACTCCAGGCGTGAGTTGCCCTTATCCAATTGACTAACCACTAGGCTTTGCAGCCCGTACAGCCACAGACAAAGGCTGTTTCCCTCAGATACCCGTTGGGTTAAGAGTGTAGACGAGAGCTACTTTTAGCCTGTGGCCAGAGGTGTGAGGCCAGCTTTAGCCTTCCGGCCGGTATCTTGTAATACACACCGTCGGCGGGAAAGAATGTTCAAGATATGCTAGCGCCTTGGGCGAACT\n"
     ]
    }
   ],
   "source": [
    "s = input()\n",
    "sc = ''\n",
    "complements = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "for symb in s:\n",
    "    sc += complements[symb]\n",
    "print(sc[::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rabbits and Recurrence Relations\n",
    "\n",
    "Given: Positive integers n≤40 and k≤5.\n",
    "\n",
    "Return: The total number of rabbit pairs that will be present after n months, if we begin with 1 pair and in each generation, every pair of reproduction-age rabbits produces a litter of k rabbit pairs (instead of only 1 pair)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30 4\n",
      "436390025825\n"
     ]
    }
   ],
   "source": [
    "n, k = (int(i) for i in input().split())\n",
    "\n",
    "def Fib(n,k):\n",
    "    if n == 1:\n",
    "        return 1\n",
    "    elif n == 2:\n",
    "        return 1\n",
    "    else:\n",
    "        return Fib(n-1, k) + k*Fib(n-2, k)\n",
    "    \n",
    "print(Fib(n,k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Computing GC Content\n",
    "\n",
    "Given: At most 10 DNA strings in FASTA format (of length at most 1 kbp each).\n",
    "\n",
    "Return: The ID of the string having the highest GC-content, followed by the GC-content of that string. Rosalind allows for a default error of 0.001 in all decimal answers unless otherwise stated; please see the note on absolute error below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Rosalind_7829': 52.9351184346035, 'Rosalind_2571': 50.73170731707317, 'Rosalind_9008': 50.76023391812865, 'Rosalind_7775': 49.44512946979038, 'Rosalind_4806': 50.58139534883721, 'Rosalind_3317': 50.70270270270271, 'Rosalind_2337': 51.97568389057751, 'Rosalind_6037': 49.2063492063492, 'Rosalind_3785': 51.64271047227926}\n",
      "Rosalind_7829\n",
      "52.9351184346035\n"
     ]
    }
   ],
   "source": [
    "ID_GCcontent = {}\n",
    "name = '-1'\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        if line[0] == '>':\n",
    "            if name != '-1':\n",
    "                ID_GCcontent[name] = count/lens * 100\n",
    "            name = line[1:-1]\n",
    "            count = 0\n",
    "            lens = 0            \n",
    "        else:\n",
    "            lens += len(line.strip())\n",
    "            # print('||' + line.strip() + '||')\n",
    "            count += line.count('C') + line.count('G')\n",
    "ID_GCcontent[name] = count/lens * 100\n",
    "print(ID_GCcontent)\n",
    "\n",
    "max_val = max(ID_GCcontent.values())\n",
    "for key in ID_GCcontent:\n",
    "    if ID_GCcontent[key] == max_val:\n",
    "        print(key)\n",
    "        print(max_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Counting Point Mutations\n",
    "\n",
    "Given: Two DNA strings s and t of equal length (not exceeding 1 kbp).\n",
    "\n",
    "Return: The Hamming distance dH(s,t)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CGAGCCCAATTGTGTACCCCCTTGCAGATTATAAGTGAACTCTCATACAGACGGGCTTAAATCCGCATTTTCAGCAACCAACCTGTTGCCCACTGTCCTAAGCGTCCACAGGATTTGTGACTGAAGAGAAACCCCGATGGAAGTACTCCGTTAGTGAGTCTCGCGTCGGGGGGAACAAGTCAAAATAACTTCATCGTTGTTACAGAGGATTATCTCTTCCCGGCCAAAAGTAGTGTCCAACAGCGTAAATTGCACTCTCCTCCCTGTCACGCGCCTGCTTTAGTTACGGCAGAGCTGCCCTACTGGTCTTGGTAGATACATGGATCCTTCGGCGAACAAGTGGATGCGGCCGAGCCGCTATAGCTTTTGACTCCTCAACATAAGATATAAACGCGGAGTATAGGCATCGGATCCCAGAACCGACCTCGAACTGGAAGTCCCTCGGAGAGGTAATGAGTGCTTCTTCATCATCACGAACGCAATTAACCGACGGACATGTCGAGGAGTGCATCTGCGAAGTCCGGCTTCGGTCAGAACCGTAGATGCGGTGGGTGGGAAGATATCGACTTTCTCACAGCCACGTGAAACCGGACCCAATGGAATTGTTAACACTCCAGGTAAGCACGTACAGGGGGCTCATCTCCGCTATGACTTGTTGGGATGCAATGAAATGATTTAGCACTGTATGTCAGGGATATTGGCTAGGGTTCGGACACTTCAAGCTTTGTCGGGATGCGCTGATGTGGTCATATCTATAAGACGGCGGTTGATTACGCAGGGGCCGACGTCGGCTGTAACGGTGTATCTGCCCAAATACGACACTGGGCAAACAACTTAGCTTAAGGTCGGATAGCAATTTGAGGACTAAATATCTTGATTACTTTACGGCGACAGACTTGACTAGCGGAGGAG\n",
      "TGAGTCTTCTAGGAGTGCCGATTGCCATTATTACTTCAAGTCTCTTATAGTGGGCTTTCACTCGGCAGATTAGCCAAGCAACCAAGCGCAGAGGCGTCCCCGCGGCATCCAGATTTGTGAAAGGAGGCAATAGAGCATGTTTATACCCAAGTCGTAAGTCATGCTCCCTATAGATCGAACGTAAAAACGTTCAGTGCTCATAGGCGGCTTCATGTCATCCCACCGACGATCAGAGATCATCACCGGGTCTTACGTGTGCACCCCTGTAACCCGCTTGCTTTCGTGGCTTCAGGCGGTATCTATGAGTCTGCGTCTGGATAAGAACCCTACGGCCAACTTTGTGAGGCGGCCGAACCTAGAACCCTGTTGACACCTAAATACAACGCGGTAATGCGGGGTTTAGGCACCGTAGGCCAAAGCGGTCTTGTAACTGTAAAAGTCGTCGTGCGTTATTGGGTTCTTCTCCCGACATCCGTACTAAATTCAACCGTAAATATGCCAGAGTGTGGCTGCGAATACTAGGGTGACTTGTTGGCACAATAGCCCGGGCTGTCTGATGGTATGAACATTGTAACAATATTAACTATATAGAAGCCGTGCCACTGGTGATACTGGATATATGCACACATAGGTGGCTCGGGACCGCAATCATTCGGGTACATTAAACTAATGGGCGTAAGAATCAATCCCACCGGGTTTAGCAATTGAATTGACACGTGAAGGTGCGTCGCGGAGAGCTACTATAGTGCCACCCCTGATCGGGGTAATGCAGAGGGAGCCGACGCAGTCGCTTCTGAAGGACATTCAAACCTCTTACGTTCCTGGGCAGGCAATGGCTCGTATGGGCGAATAACTGTTTTTTTTCTAAATAGCGAGATGACGCGACTCTCATGGACCGTCAATACCAATGCG\n",
      "452\n"
     ]
    }
   ],
   "source": [
    "def hamming_dist(s, t):\n",
    "    dist = 0\n",
    "    for i in range(len(s)):\n",
    "        if s[i] != t[i]:\n",
    "            dist += 1\n",
    "    return dist\n",
    "\n",
    "s = input()\n",
    "t = input()\n",
    "print(hamming_dist(s, t))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mendel's First Law\n",
    "\n",
    "Given: Three positive integers k, m, and n, representing a population containing k+m+n organisms: k individuals are homozygous dominant for a factor, m are heterozygous, and n are homozygous recessive.\n",
    "\n",
    "Return: The probability that two randomly selected mating organisms will produce an individual possessing a dominant allele (and thus displaying the dominant phenotype). Assume that any two organisms can mate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24 30 27\n",
      "0.733101851851852\n"
     ]
    }
   ],
   "source": [
    "def mendel_first_law(k,m,n):\n",
    "    kmn = k+m+n\n",
    "    k_k = 1   * k/kmn*(k-1)/(kmn-1)\n",
    "    m_m = 3/4 *(m/kmn * (m-1)/(kmn-1))\n",
    "    n_n = 0\n",
    "    \n",
    "    k_m = 1   *k/kmn*m/(kmn-1) + 1   *m/kmn*k/(kmn-1)\n",
    "    k_n = 1   *k/kmn*n/(kmn-1) + 1   *n/kmn*k/(kmn-1)\n",
    "    m_n = 1/2 *m/kmn*n/(kmn-1) + 1/2 *n/kmn*m/(kmn-1)\n",
    "    \n",
    "    sum = k_k + m_m + n_n + k_m + k_n + m_n\n",
    "    return sum\n",
    "\n",
    "k, m, n = (int(i) for i in (input().split()))\n",
    "print(mendel_first_law(k,m,n))    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Translating RNA into Protein\n",
    "\n",
    "Given: An RNA string s corresponding to a strand of mRNA (of length at most 10 kbp).\n",
    "\n",
    "Return: The protein string encoded by s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate(seq): \n",
    "       \n",
    "    table = {\"UUU\":\"F\", \"UUC\":\"F\", \"UUA\":\"L\", \"UUG\":\"L\",\n",
    "    \"UCU\":\"S\", \"UCC\":\"S\", \"UCA\":\"S\", \"UCG\":\"S\",\n",
    "    \"UAU\":\"Y\", \"UAC\":\"Y\", \"UAA\":\"STOP\", \"UAG\":\"STOP\",\n",
    "    \"UGU\":\"C\", \"UGC\":\"C\", \"UGA\":\"STOP\", \"UGG\":\"W\",\n",
    "    \"CUU\":\"L\", \"CUC\":\"L\", \"CUA\":\"L\", \"CUG\":\"L\",\n",
    "    \"CCU\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "    \"CAU\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "    \"CGU\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "    \"AUU\":\"I\", \"AUC\":\"I\", \"AUA\":\"I\", \"AUG\":\"M\",\n",
    "    \"ACU\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "    \"AAU\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "    \"AGU\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "    \"GUU\":\"V\", \"GUC\":\"V\", \"GUA\":\"V\", \"GUG\":\"V\",\n",
    "    \"GCU\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "    \"GAU\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "    \"GGU\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"} \n",
    "    protein = ''\n",
    "    for i in range(0, len(seq), 3):\n",
    "        RNA_codon = seq[i:i+3]\n",
    "        if table[RNA_codon] != 'STOP':\n",
    "            protein += table[RNA_codon]\n",
    "        else:\n",
    "            #protein += '\\n'\n",
    "            break\n",
    "    return protein\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    seq = inf.read().strip()\n",
    "\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    ouf.write(translate(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a Motif in DNA\n",
    "\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp).\n",
    "\n",
    "Return: All locations of t as a substring of s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GGGACGGAAAGACGGAACTAGTGTCAAAGGACGGAAAGACGGAATGACGGAAGTTCCGACGGAAGACGGAAGCGACGGAATTGACGGAAGGGACGGAATACCGTTATGGGCGACGGAAGGACGGAATAGACGGAAGACGGAAGACGGAAACCTGACGGAAGGCCCACGAGGACGGAAGACGGAATTAGAGACGGAATGACGGAAGGACGGAAGACGGAATCGGACGGAACACGACGGAAGGACGGAAGACGGAAGGAGCGACGGAACCCGACGGAAGACGGAAACGATGGCTCGCGACGGAAGCGACGGAAAGACGGAACCCGACGGAAGACGGAAGCAGGACGGAATTACGGGACGGAACAATGGGACGGAAGGACGGAAATGACGGAAGCCCGTCGACGGAATGTGCGCTTTCGACGGAAAACTATGACGGAAGACGGAACGACGGAACCCGGACGGAACGACGGAATTGACGGAAGGTCGGACGGAAGGACGGAATGACGGAAGGACGGAAAGACGGAACCGGACGGAACGACGGAAGACGGAAGACGGAAGACGGAAGAGACGGAAGACGGAAGACGGAACGACGGAAACCCGACGGAAATACTCTTCGACGGAAGACGGAATTCGACGGAAGGACGGAAAGACGGAAGACGGAAAGACTTTCCGAGACGGAAGACGGAAACCGACGGAAATGTGACGGAAGGACGGAAGACGGAATGACGGAATCCGACGGAACACCCGCGACGGAAGACGGAAGACGGAACGACGGAAGACGGAACATGACGGAAGACGGAAAGGACGGAAGAGACGGAAAGACGGAAGACGGAAGACCGCCTTCTTGACGGAAAGTGCGACGGAATGACGGAAGCGACGGAACGTCCCGACGACGGAAAAGACGGAA\n",
      "GACGGAAGA\n",
      "58 129 136 171 206 241 270 323 429 534 541 548 555 564 571 613 646 671 707 746 753 768 785 801 818 825\n"
     ]
    }
   ],
   "source": [
    "s = input()\n",
    "t = input()\n",
    "i = 0\n",
    "starts = []\n",
    "len_s = len(s)\n",
    "len_t = len(t)\n",
    "while i < len_s+1:\n",
    "    pos = s[i:].find(t)\n",
    "    if pos != -1:\n",
    "        starts.append(pos+i+1)\n",
    "        i += pos+1\n",
    "    else: \n",
    "        break\n",
    "print(*starts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Consensus and Profile\n",
    "\n",
    "Given: A collection of at most 10 DNA strings of equal length (at most 1 kbp) in FASTA format.\n",
    "\n",
    "Return: A consensus string and profile matrix for the collection. (If several possible consensus strings exist, then you may return any one of them.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read DNA strings from file\n",
    "\n",
    "DNA_strings = []\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    reading_string = False\n",
    "    string = ''\n",
    "    while True:\n",
    "        line = inf.readline().strip()\n",
    "        if (not line) or line[0] == '>':\n",
    "            if reading_string:\n",
    "                DNA_strings.append(string)\n",
    "                string = ''\n",
    "            else:\n",
    "                reading_string = True\n",
    "            if (not line):\n",
    "                break\n",
    "        else:\n",
    "            string += line\n",
    "        \n",
    "                \n",
    "    \n",
    "#     for line in inf:\n",
    "#         if (line[0] == '>') :\n",
    "#             DNA_strings.append(string)\n",
    "#             string = ''\n",
    "#         elif (line[0] != '>'):\n",
    "#             string += line\n",
    "\n",
    "\n",
    "# for s in DNA_strings:\n",
    "#     print(len(s))\n",
    "\n",
    "#calculate profile matrix\n",
    "\n",
    "def profile_matrix(DNA_strings):\n",
    "    n = len(DNA_strings[0])\n",
    "    labels = ['A', 'C', 'G', 'T']\n",
    "    matr =  [[0] * n for i in range(4)]\n",
    "    for i,label in enumerate(labels):\n",
    "        for j in range(n):\n",
    "            count = 0\n",
    "            for k,string in enumerate(DNA_strings):\n",
    "                if string[j] == label:\n",
    "                     count += 1\n",
    "            matr[i][j] = count\n",
    "    return matr\n",
    "\n",
    "\n",
    "pr_matrix = profile_matrix(DNA_strings)\n",
    "\n",
    "#calculate consensus string\n",
    "\n",
    "def consensus_string(profile_matrix):\n",
    "    numrows = len(profile_matrix)\n",
    "    numcols = len(profile_matrix[0])\n",
    "    labels = ['A', 'C', 'G', 'T']\n",
    "    cons_str = [0 for i in range(numcols)]\n",
    "    for j in range(numcols):\n",
    "        max_el, ind = -1, -1\n",
    "        for i in range(numrows):\n",
    "            el = profile_matrix[i][j]\n",
    "            if el > max_el:\n",
    "                max_el, ind = el, i\n",
    "        cons_str[j] = labels[ind]\n",
    "    return cons_str\n",
    "\n",
    "cons_string = consensus_string(pr_matrix)\n",
    "\n",
    "#write in file\n",
    "\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "\n",
    "    for el in cons_string:\n",
    "        ouf.write(el)\n",
    "    ouf.write('\\n')\n",
    "    \n",
    "    numrows = len(pr_matrix)\n",
    "    numcols = len(pr_matrix[0])\n",
    "    labels = ['A', 'C', 'G', 'T']\n",
    "    \n",
    "    for i in range(numrows):\n",
    "        ouf.write(labels[i]+': ')\n",
    "        for j in range(numcols):\n",
    "            ouf.write(str(pr_matrix[i][j])+' ')\n",
    "        ouf.write('\\n')\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mortal Fibonacci Rabbits\n",
    "\n",
    "Given: Positive integers n≤100 and m≤20.\n",
    "\n",
    "Return: The total number of pairs of rabbits that will remain after the n-th month if all rabbits live for m months."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93 20\n",
      "[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0]\n",
      "[144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0]\n",
      "[233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0, 0]\n",
      "[377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0, 0]\n",
      "[610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0, 0]\n",
      "[987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0, 0]\n",
      "[1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1, 0]\n",
      "[2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0, 1]\n",
      "[4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1, 0]\n",
      "[6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1, 1]\n",
      "[10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2, 1]\n",
      "[17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3, 2]\n",
      "[28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5, 3]\n",
      "[46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8, 5]\n",
      "[75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13, 8]\n",
      "[121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21, 13]\n",
      "[196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34, 21]\n",
      "[317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55, 34]\n",
      "[514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89, 55]\n",
      "[831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144, 89]\n",
      "[1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233, 144]\n",
      "[2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377, 233]\n",
      "[3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610, 377]\n",
      "[5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987, 610]\n",
      "[9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597, 987]\n",
      "[14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584, 1597]\n",
      "[24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181, 2584]\n",
      "[39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764, 4181]\n",
      "[63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945, 6764]\n",
      "[102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708, 10945]\n",
      "[165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652, 17708]\n",
      "[267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358, 28652]\n",
      "[433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007, 46358]\n",
      "[700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360, 75007]\n",
      "[1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359, 121360]\n",
      "[1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706, 196359]\n",
      "[2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044, 317706]\n",
      "[4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716, 514044]\n",
      "[7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705, 831716]\n",
      "[12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332, 1345705]\n",
      "[20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893, 2177332]\n",
      "[32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992, 3522893]\n",
      "[53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508, 5699992]\n",
      "[86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890, 9222508]\n",
      "[139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411, 14921890]\n",
      "[225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704, 24143411]\n",
      "[365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531, 39063704]\n",
      "[590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054, 63204531]\n",
      "[955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821, 102264054]\n",
      "[1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930, 165461821]\n",
      "[2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043, 267714930]\n",
      "[4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321, 433159043]\n",
      "[6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006, 700845321]\n",
      "[10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320, 1133958006]\n",
      "[17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966, 1834728320]\n",
      "[27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927, 2968564966]\n",
      "[44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187, 4803096927]\n",
      "[72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070, 7771344187]\n",
      "[117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541, 12573927070]\n",
      "[190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906, 20344439541]\n",
      "[307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115, 32917020906]\n",
      "[497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128, 53259283115]\n",
      "[805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251, 86172781128]\n",
      "[1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871, 139426364251]\n",
      "[2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232, 225589922871]\n",
      "[3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692, 365001365232]\n",
      "[5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220, 590567144692]\n",
      "[8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381, 955529446220]\n",
      "[14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547, 1546033386381]\n",
      "[23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107, 2501460568547]\n",
      "[37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724, 4047328493107]\n",
      "[61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788, 6548521346724]\n",
      "[99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191, 10595416680788]\n",
      "[160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973, 17143237182191]\n",
      "[259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844, 27737519904973]\n",
      "[419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851, 44878922358844]\n",
      "[678481924848083841, 419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768, 72613473698851]\n",
      "[1097774340728876973, 678481924848083841, 419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432, 117487592960768]\n",
      "[1776183652103261963, 1097774340728876973, 678481924848083841, 419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130, 190093295315432]\n",
      "[2873840505239178168, 1776183652103261963, 1097774340728876973, 678481924848083841, 419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021, 307568314349130]\n",
      "[4649834064047124699, 2873840505239178168, 1776183652103261963, 1097774340728876973, 678481924848083841, 419337294803151976, 259172367564836838, 160182070475497329, 99000892506020297, 61187726490823756, 37817213343689648, 23373014607702655, 14445744769373374, 8928225367775501, 5518109968742565, 3410480400398168, 2107855158267268, 1302764668495151, 805176662553245, 497641265225021]\n",
      "12173201065019078436\n"
     ]
    }
   ],
   "source": [
    "n, m = (int(i) for i in input().split())\n",
    "\n",
    "def one_month_later(rabits):\n",
    "    s_new = sum(rabits[1:])\n",
    "    for i in reversed(range(1,m)):\n",
    "        rabits[i] = rabits[i-1]\n",
    "    rabits[0] = s_new\n",
    "\n",
    "rabits = [0]*m\n",
    "rabits[0] = 1\n",
    "\n",
    "print(rabits)\n",
    "\n",
    "for i in range(1,n):\n",
    "    one_month_later(rabits)\n",
    "    print(rabits)\n",
    "    \n",
    "print(sum(rabits))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overlap Graphs\n",
    "\n",
    "Given: A collection of DNA strings in FASTA format having total length at most 10 kbp.\n",
    "\n",
    "Return: The adjacency list corresponding to O3. You may return edges in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_DNA = {}\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            name = line[1:]\n",
    "            name_DNA[name] = ''\n",
    "        else:\n",
    "            name_DNA[name] += line\n",
    "\n",
    "adjacency_list = []\n",
    "            \n",
    "for key, val in name_DNA.items():\n",
    "    for key_next, val_next in name_DNA.items():\n",
    "        if val != val_next:\n",
    "            if val[-3:] == val_next[:3]:\n",
    "                adjacency_list.append([key, key_next])\n",
    "                \n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    for el in adjacency_list:\n",
    "        ouf.write(' '.join(el))\n",
    "        ouf.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculating Expected Offspring\n",
    "\n",
    "Given: Six nonnegative integers, each of which does not exceed 20,000. The integers correspond to the number of couples in a population possessing each genotype pairing for a given factor. In order, the six given integers represent the number of couples having the following genotypes:\n",
    "\n",
    "AA-AA\n",
    "AA-Aa\n",
    "AA-aa\n",
    "Aa-Aa\n",
    "Aa-aa\n",
    "aa-aa\n",
    "\n",
    "Return: The expected number of offspring displaying the dominant phenotype in the next generation, under the assumption that every couple has exactly two offspring."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19462 19541 19205 18955 16280 17919\n",
      "161128.5\n"
     ]
    }
   ],
   "source": [
    "probability = {'AA-AA': 1, 'AA-Aa': 1, 'AA-aa': 1, 'Aa-Aa': 0.75, 'Aa-aa': 0.5, 'aa-aa':0}\n",
    "parents_genotype = ['AA-AA', 'AA-Aa', 'AA-aa', 'Aa-Aa', 'Aa-aa', 'aa-aa']\n",
    "parents = list(int(i) for i in input().split())\n",
    "\n",
    "expectation = 0\n",
    "for i,parent_pair in enumerate(parents):\n",
    "    expectation += probability[parents_genotype[i]]*(parent_pair)*2\n",
    "    \n",
    "print(expectation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a Shared Motif\n",
    "\n",
    "Given: A collection of k (k≤100) DNA strings of length at most 1 kbp each in FASTA format.\n",
    "\n",
    "Return: A longest common substring of the collection. (If multiple solutions exist, you may return any single solution.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCATCATATGGCCCGTATGCGTTGGAACTGCGGGATCCGGGGCGCTAATGAGCGGGGGGATGACAAAGACATTCAGTGACGTGGCTACATTGATCTTCCTCGGACCCTCCGACAGTATATTCCTGACCGAGGAACTGGCTT\n"
     ]
    }
   ],
   "source": [
    "name_DNA = {}\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            name = line[1:]\n",
    "            name_DNA[name] = ''\n",
    "        else:\n",
    "            name_DNA[name] += line\n",
    "            \n",
    "DNA = list(name_DNA.values())\n",
    "DNA.sort(key=len)\n",
    "\n",
    "substrings = []\n",
    "com_sub = DNA[0]\n",
    "i = 0\n",
    "k = 0\n",
    "try_this = com_sub[i]\n",
    "while (k) < len(com_sub):\n",
    "    for string in DNA:\n",
    "        if try_this not in string:\n",
    "            i += 1\n",
    "            k = i\n",
    "            if i < len(com_sub):\n",
    "                try_this = com_sub[i]\n",
    "            break\n",
    "    else:\n",
    "        substrings.append(try_this)\n",
    "        k += 1\n",
    "        if i < len(com_sub):\n",
    "            try_this = com_sub[i:k]\n",
    "            \n",
    "\n",
    "substrings.sort(key=len)\n",
    "print(substrings[-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Independent Alleles\n",
    "\n",
    "Given: Two positive integers k (k≤7) and N (N≤2k). In this problem, we begin with Tom, who in the 0th generation has genotype Aa Bb. Tom has two children in the 1st generation, each of whom has two children, and so on. Each organism always mates with an organism having genotype Aa Bb.\n",
    "\n",
    "Return: The probability that at least N Aa Bb organisms will belong to the k-th generation of Tom's family tree (don't count the Aa Bb mates at each level). Assume that Mendel's second law holds for the factors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 34\n",
      "0.37394183862382063\n"
     ]
    }
   ],
   "source": [
    "k, N = (int(i) for i in input().split())\n",
    "\n",
    "import math\n",
    "\n",
    "def p_n(n, k):\n",
    "    C_2k_n = (math.factorial(2**k)/(math.factorial(n)*math.factorial((2**k)-n)))\n",
    "    P_n = (C_2k_n*(0.25**n)*(0.75**(2**k - n)))\n",
    "    return P_n\n",
    "\n",
    "probability = 0\n",
    "for i in range(N, 2**k+1):\n",
    "    probability += p_n(i, k)\n",
    "    \n",
    "print(probability)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a Protein Motif\n",
    "\n",
    "Given: At most 15 UniProt Protein Database access IDs.\n",
    "\n",
    "Return: For each protein possessing the N-glycosylation motif N{P}[ST]{P}, output its given access ID followed by a list of locations in the protein string where the motif can be found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import re\n",
    "\n",
    "def get_a_string(new_link):\n",
    "    page = requests.get(new_link)\n",
    "    file = page.text\n",
    "    string = file.strip().split('\\n')\n",
    "    new_string = ''\n",
    "    for el in string[1:]:\n",
    "        new_string += el\n",
    "    return new_string\n",
    "\n",
    "def find_a_pattern(new_string, pattern):\n",
    "    pos = []\n",
    "    k = 0\n",
    "    while k < len(new_string):\n",
    "        m = re.search(pattern, new_string[k:])\n",
    "        if m:\n",
    "            place = m.span()[0] + 1\n",
    "            k += place\n",
    "            pos.append(k)\n",
    "        else:\n",
    "            break\n",
    "    return pos\n",
    "        \n",
    "name_loc = {}\n",
    "pattern = r'N[^P][ST][^P]'\n",
    "        \n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        new_link = 'http://www.uniprot.org/uniprot/' + line.strip() + '.fasta'\n",
    "        new_string = get_a_string(new_link)\n",
    "        positions = find_a_pattern(new_string, pattern)\n",
    "        name_loc[line.strip()] = positions      \n",
    "\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    for key in name_loc:\n",
    "        if len(name_loc[key]) != 0:\n",
    "            ouf.write(key)\n",
    "            ouf.write('\\n')\n",
    "            ouf.write(' '.join(str(i) for i in name_loc[key]))\n",
    "            ouf.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inferring mRNA from Protein\n",
    "\n",
    "Given: A protein string of length at most 1000 aa.\n",
    "\n",
    "Return: The total number of different RNA strings from which the protein could have been translated, modulo 1,000,000. (Don't neglect the importance of the stop codon in protein translation.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MWLVYNQYCKLRDKAGHNKRFDFYQEQVEVEDWICEMSRERLWNGFWQEVLHIYFGYMYFEGIDYDWHNILHTRRWQTSRQQPDRKGWQLVGSRSCPIAWTNMIQFDLVEQPQIWGAEAANMMAEFWSRSQMAPRLSDWAHNQSFAASLPMSLHFAPIMEHDIVVWCQIVHNPHKPWMCYSIHQWQYADMYWAFPQCPFICPTNAPCKLMLNEINGYISKGMAPTSWCTAFWAMKRTPWFMSTFKRHKWMQYVDWVPMPSNRCCVCRQRFGKFSQTYPQWMEWLHPKAKYGRGTVMMQGWCMNYDLWMKNHKPNFCLFRIHCMYNVGDGGVWPICTMALWFMFWITTTNSQWRTHGMQPWAYFGSCDMDDQDIVWNQNTNLIGLFYNVHGQGRQWTIPICWSFRNPPDHTQQWEDGFGEQVLMASVLYNVLYNNTIHAHEVGNQQSANFMVEPRHPSFIEGEFFVIDLYEFTASVMLNCSEVCWRTGEEVTPVQWSMKMAPSICGASWDWYPFYVNWFACFQLIAFHIGLGAESPIPQHHRIFCNMWSRFGSWQGNKPMTFYQMTGGHNSPFFINGLKGIKMVGWIRHTMCHGWIFVWDGISMMYYTGSWMEFVDYMMNEIDMESYMCFPKRSNETACGCNRGHESIMVVPGWQLQKCACCFLCLLCYRGYWAHAIPCVTIPEMHQMDPNKHENEDQKAGAGRTVTIAQPLKFHPSIHDDPNEAFPDRMVFLTDDSKLPCVLWSACANVMGICITIMHYLCLPRGAHTILMCNHWIVHDPETFKGIEYIMPASYASFFEALFMYVQFDQWIMKYMNNDIKQTHRFERQQFWHARMMETWRKHNQSLSPRQYNVKNHVLDWFYIDPFACYITTKAEKHDSQYEWEEDFKNFPPPFDQSGQCRGLRRKNIEIVQYITSDKRFRHDTRKAGLNGMMLLSQCNFSLPHFECQIRQRFIFMEKKYMKACSDWRPGMQPTDIGNHHMGPICKDGYMPSKDNGIGH\n",
      "\n",
      "\n",
      "749568\n"
     ]
    }
   ],
   "source": [
    "table_RNA_to_protein = {\"UUU\":\"F\", \"UUC\":\"F\", \"UUA\":\"L\", \"UUG\":\"L\",\n",
    "\"UCU\":\"S\", \"UCC\":\"S\", \"UCA\":\"S\", \"UCG\":\"S\",\n",
    "\"UAU\":\"Y\", \"UAC\":\"Y\", \"UAA\":\"Stop\", \"UAG\":\"Stop\",\n",
    "\"UGU\":\"C\", \"UGC\":\"C\", \"UGA\":\"Stop\", \"UGG\":\"W\",\n",
    "\"CUU\":\"L\", \"CUC\":\"L\", \"CUA\":\"L\", \"CUG\":\"L\",\n",
    "\"CCU\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "\"CAU\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "\"CGU\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "\"AUU\":\"I\", \"AUC\":\"I\", \"AUA\":\"I\", \"AUG\":\"M\",\n",
    "\"ACU\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "\"AAU\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "\"AGU\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "\"GUU\":\"V\", \"GUC\":\"V\", \"GUA\":\"V\", \"GUG\":\"V\",\n",
    "\"GCU\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "\"GAU\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "\"GGU\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\"}\n",
    "\n",
    "protein = list(''.join(input().strip()))\n",
    "protein.append('Stop')\n",
    "\n",
    "combinations = 1\n",
    "\n",
    "for el in protein:\n",
    "    count = 0\n",
    "    for key, value in table_RNA_to_protein.items():\n",
    "        if value == el:\n",
    "            count += 1\n",
    "    combinations *= count\n",
    "\n",
    "print(combinations % 1000000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Open Reading Frames\n",
    "\n",
    "Given: A DNA string s of length at most 1 kbp in FASTA format.\n",
    "\n",
    "Return: Every distinct candidate protein string that can be translated from ORFs of s. Strings can be returned in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_codon_table = {\n",
    "'TTT': 'F',      'CTT': 'L',      'ATT': 'I',      'GTT': 'V',\n",
    "'TTC': 'F',      'CTC': 'L',      'ATC': 'I',      'GTC': 'V',\n",
    "'TTA': 'L',      'CTA': 'L',      'ATA': 'I',      'GTA': 'V',\n",
    "'TTG': 'L',      'CTG': 'L',      'ATG': 'M',      'GTG': 'V',\n",
    "'TCT': 'S',      'CCT': 'P',      'ACT': 'T',      'GCT': 'A',\n",
    "'TCC': 'S',      'CCC': 'P',      'ACC': 'T',      'GCC': 'A',\n",
    "'TCA': 'S',      'CCA': 'P',      'ACA': 'T',      'GCA': 'A',\n",
    "'TCG': 'S',      'CCG': 'P',      'ACG': 'T',      'GCG': 'A',\n",
    "'TAT': 'Y',      'CAT': 'H',      'AAT': 'N',      'GAT': 'D',\n",
    "'TAC': 'Y',      'CAC': 'H',      'AAC': 'N',      'GAC': 'D',\n",
    "'TAA': 'Stop',   'CAA': 'Q',      'AAA': 'K',      'GAA': 'E',\n",
    "'TAG': 'Stop',   'CAG': 'Q',      'AAG': 'K',      'GAG': 'E',\n",
    "'TGT': 'C',      'CGT': 'R',      'AGT': 'S',      'GGT': 'G',\n",
    "'TGC': 'C',      'CGC': 'R',      'AGC': 'S',      'GGC': 'G',\n",
    "'TGA': 'Stop',   'CGA': 'R',      'AGA': 'R',      'GGA': 'G',\n",
    "'TGG': 'W',      'CGG': 'R',      'AGG': 'R',      'GGG': 'G'\n",
    "}\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    s_DNA = ''\n",
    "    for line in inf:\n",
    "        if line[0] != '>':\n",
    "            s_DNA += line.strip()\n",
    "\n",
    "start_DNA = 'ATG'\n",
    "stop_DNA = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "\n",
    "DNA_seq = []\n",
    "def make_DNA_sequences(s_DNA):\n",
    "\n",
    "    try_start = []\n",
    "    try_stop = []\n",
    "\n",
    "    i = 0\n",
    "    while i < len(s_DNA)-3:\n",
    "        if s_DNA[i:i+3] == start_DNA:\n",
    "            try_start.append(i)\n",
    "        for el in stop_DNA:\n",
    "            if s_DNA[i:i+3] == el:\n",
    "                try_stop.append(i)\n",
    "        i += 1\n",
    "\n",
    "    \n",
    "    for start_loc in try_start:\n",
    "        for stop_loc in try_stop:\n",
    "            if stop_loc > start_loc and (stop_loc-start_loc) % 3 == 0:\n",
    "                DNA_seq.append(s_DNA[start_loc:stop_loc])\n",
    "                break\n",
    "\n",
    "            \n",
    "def translate_DNA_to_protein(DNA_seq):\n",
    "    protein = ''\n",
    "    for i in range(0, len(DNA_seq), 3):\n",
    "        cod = DNA_seq[i:i+3]\n",
    "        protein += DNA_codon_table[cod]\n",
    "    return protein\n",
    "        \n",
    "    \n",
    "complements = {\n",
    "    'A' : 'T',\n",
    "    'T' : 'A',\n",
    "    'C' : 'G',\n",
    "    'G' : 'C'\n",
    "}\n",
    "def reverse_complement(dna):\n",
    "    dna_new = ''\n",
    "    for i in reversed(range(len(dna))):\n",
    "        dna_new += complements[dna[i]]\n",
    "    return dna_new\n",
    "\n",
    "                \n",
    "\n",
    "make_DNA_sequences(s_DNA)\n",
    "dna_reversed = reverse_complement(s_DNA)\n",
    "make_DNA_sequences(dna_reversed)\n",
    "\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    for seq in set(DNA_seq):\n",
    "        protein = translate_DNA_to_protein(seq)\n",
    "        ouf.write(protein)\n",
    "        ouf.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enumerating Gene Orders\n",
    "\n",
    "Given: A positive integer n≤7.\n",
    "\n",
    "Return: The total number of permutations of length n, followed by a list of all such permutations (in any order)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "6\n",
      "1 2 3\n",
      "1 3 2\n",
      "2 1 3\n",
      "2 3 1\n",
      "3 1 2\n",
      "3 2 1\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "import itertools\n",
    "\n",
    "n = int(input())\n",
    "print(math.factorial(n))\n",
    "perm = list(itertools.permutations(list(range(1, n+1))))\n",
    "for element in perm:\n",
    "    print(*element)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculating Protein Mass\n",
    "\n",
    "Given: A protein string P of length at most 1000 aa.\n",
    "\n",
    "Return: The total weight of P. Consult the monoisotopic mass table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GHVHFIQDFSHHWTTWDTDCYAHPLWFYHITWQDNCCGPMEDACDDLCESEMEFNRDAIIVTSAGEVSKDSSSYLQLETWVNITHIIQELPLRVGCSMQLRPDKKNADMPWVFLGMKFRRNAIIGPSKLFPEYEKFIFQLLIVALPIVHWSAYYPLSINDPGEIVHKMLKYLYMVMDWCHYARVCHDPLRSAQPPHRASGGLFGLARERNSDPVQAYTFGPHFCEAPSGTAAYQLCHDKFDNMIMGGIATLTYSLQNCCVTHWTFRYVASCQTTQVVHETKCDISVIHIIGHQSYPPYKFTGKHMALSVMNAAICSECTANMDSNWLGEPCNSELLDFQTDHEWFGANPTCTWVNQQDFGECLWFWVAQCWYIFSDMLHQIYGPAFHVSILHQLFNETNEPANTRDWWSKKLYHTDSYPPFDKVQFAPNSFHQDFIEVEFTLPDMIVRGCWSTHTTNPYYQYQQFLQMWAMSECVFTPIQEYHRTGYFSSCCSEASMCLGVMDRYNVDFNLHCWLFSRCEANRTAWCVWKMKVIGTASFGHVTDLYFGNKSASKDNNSCMELRMREKLRKSGIDDLVGFIMSMVYYVEAKMMGKSMELPDDMITQPAHLFRTEENYWTEQCKQCPSAENRTPQYYADHMAQCRRVIAFQGEGWWRVGMGPWIDTYLNSTSGYYRYTTSDWFGPVASGHMFPTQSSRWVCTGIKFVCMMIVNPQVVVCLTVGFRDWRGDNCYYWGHEKHTHKAFYQEGSFCIKMRQLTDWAICFWCPHHHGGPYFVAGMDWMKNDTSKYPNVIWDWQHTVTRQYLGSCCQMVP\n",
      "94645.96185000049\n"
     ]
    }
   ],
   "source": [
    "monoisotopic_mass_table = {\n",
    "'A':   71.03711,\n",
    "'C':   103.00919,\n",
    "'D':   115.02694,\n",
    "'E':   129.04259,\n",
    "'F':   147.06841,\n",
    "'G':   57.02146,\n",
    "'H':   137.05891,\n",
    "'I':   113.08406,\n",
    "'K':   128.09496,\n",
    "'L':   113.08406,\n",
    "'M':   131.04049,\n",
    "'N':   114.04293,\n",
    "'P':   97.05276,\n",
    "'Q':   128.05858,\n",
    "'R':   156.10111,\n",
    "'S':   87.03203,\n",
    "'T':   101.04768,\n",
    "'V':   99.06841,\n",
    "'W':   186.07931,\n",
    "'Y':   163.06333 \n",
    "}\n",
    "\n",
    "protein = input()\n",
    "weight = 0\n",
    "for symb in protein:\n",
    "    weight += monoisotopic_mass_table[symb]\n",
    "print(weight)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Locating Restriction Sites\n",
    "\n",
    "Given: A DNA string of length at most 1 kbp in FASTA format.\n",
    "\n",
    "Return: The position and length of every reverse palindrome in the string having length between 4 and 12. You may return these pairs in any order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('inf.txt') as inf:\n",
    "    DNA = ''\n",
    "    for line in inf:\n",
    "        if line[0] != '>':\n",
    "            DNA += line.strip()\n",
    "            \n",
    "complements = {\n",
    "    'A' : 'T',\n",
    "    'T' : 'A',\n",
    "    'C' : 'G',\n",
    "    'G' : 'C'\n",
    "}\n",
    "            \n",
    "def reverse_complement(dna):\n",
    "    dna_new = ''\n",
    "    for i in reversed(range(len(dna))):\n",
    "        dna_new += complements[dna[i]]\n",
    "    return dna_new\n",
    "\n",
    "def if_reverse_palindrome(dna):\n",
    "    return reverse_complement(dna) == dna\n",
    "\n",
    "\n",
    "min_len = 4\n",
    "max_len = 12\n",
    "\n",
    "with open('ouf.txt','w') as ouf:\n",
    "    i = 1\n",
    "    while i < len(DNA)+1:\n",
    "        for j in range(min_len, max_len+1):\n",
    "            if i+j > len(DNA)+1:\n",
    "                break\n",
    "            if if_reverse_palindrome(DNA[i-1:i+j-1]):\n",
    "                ouf.write(str(i) + ' ' + str(j) + '\\n')\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNA Splicing\n",
    "\n",
    "Given: A DNA string s (of length at most 1 kbp) and a collection of substrings of s acting as introns. All strings are given in FASTA format.\n",
    "\n",
    "Return: A protein string resulting from transcribing and translating the exons of s. (Note: Only one solution will exist for the dataset provided.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_codon_table = {\n",
    "'TTT': 'F',      'CTT': 'L',      'ATT': 'I',      'GTT': 'V',\n",
    "'TTC': 'F',      'CTC': 'L',      'ATC': 'I',      'GTC': 'V',\n",
    "'TTA': 'L',      'CTA': 'L',      'ATA': 'I',      'GTA': 'V',\n",
    "'TTG': 'L',      'CTG': 'L',      'ATG': 'M',      'GTG': 'V',\n",
    "'TCT': 'S',      'CCT': 'P',      'ACT': 'T',      'GCT': 'A',\n",
    "'TCC': 'S',      'CCC': 'P',      'ACC': 'T',      'GCC': 'A',\n",
    "'TCA': 'S',      'CCA': 'P',      'ACA': 'T',      'GCA': 'A',\n",
    "'TCG': 'S',      'CCG': 'P',      'ACG': 'T',      'GCG': 'A',\n",
    "'TAT': 'Y',      'CAT': 'H',      'AAT': 'N',      'GAT': 'D',\n",
    "'TAC': 'Y',      'CAC': 'H',      'AAC': 'N',      'GAC': 'D',\n",
    "'TAA': 'Stop',   'CAA': 'Q',      'AAA': 'K',      'GAA': 'E',\n",
    "'TAG': 'Stop',   'CAG': 'Q',      'AAG': 'K',      'GAG': 'E',\n",
    "'TGT': 'C',      'CGT': 'R',      'AGT': 'S',      'GGT': 'G',\n",
    "'TGC': 'C',      'CGC': 'R',      'AGC': 'S',      'GGC': 'G',\n",
    "'TGA': 'Stop',   'CGA': 'R',      'AGA': 'R',      'GGA': 'G',\n",
    "'TGG': 'W',      'CGG': 'R',      'AGG': 'R',      'GGG': 'G'\n",
    "}\n",
    "\n",
    "def translate_DNA_to_protein(DNA_seq):\n",
    "    protein = ''\n",
    "    for i in range(0, len(DNA_seq), 3):\n",
    "        cod = DNA_seq[i:i+3]\n",
    "        protein += DNA_codon_table[cod]\n",
    "    return protein\n",
    "\n",
    "name_DNA = {}\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            name = line[1:]\n",
    "            name_DNA[name] = ''\n",
    "        else:\n",
    "            name_DNA[name] += line\n",
    "            \n",
    "dna_introns = list(name_DNA.values())\n",
    "dna = dna_introns[0]\n",
    "introns = dna_introns[1:]\n",
    "\n",
    "for intron in introns:\n",
    "    dna = dna.replace(intron, '')\n",
    "    \n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    ouf.write(translate_DNA_to_protein(dna)[:-4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enumerating k-mers Lexicographically\n",
    "\n",
    "Given: A collection of at most 10 symbols defining an ordered alphabet, and a positive integer n (n≤10).\n",
    "\n",
    "Return: All strings of length n that can be formed from the alphabet, ordered lexicographically (use the standard order of symbols in the English alphabet)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A B C D\n",
      "4\n"
     ]
    }
   ],
   "source": [
    "letters = list(input().strip().split())\n",
    "n = int(input())\n",
    "\n",
    "import itertools\n",
    "\n",
    "permutations = set(itertools.permutations(letters*n, n))\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    for element in sorted(permutations):\n",
    "        for i in range(len(element)):\n",
    "            ouf.write(element[i])\n",
    "        ouf.write('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Longest Increasing Subsequence\n",
    "\n",
    "Given: A positive integer n≤10000 followed by a permutation π of length n.\n",
    "\n",
    "Return: A longest increasing subsequence of π, followed by a longest decreasing subsequence of π."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # TOOO slow approach\n",
    "\n",
    "\n",
    "# with open('inf.txt') as inf:\n",
    "#     n = inf.readline()\n",
    "#     permutation = []\n",
    "#     while True:\n",
    "#         my_list = list(inf.readline().strip().split())\n",
    "#         permutation.extend(my_list)\n",
    "#         if len(my_list) == 0:\n",
    "#             break\n",
    "  \n",
    "# print('find the longest increasing subsequence')\n",
    "# # find the longest increasing subsequence\n",
    "# incr_subseq = []\n",
    "# for i in range(len(permutation)):\n",
    "#     print(i)    \n",
    "#     if i+1 == (len(permutation)):\n",
    "#         incr_subseq.append(current_subseq)\n",
    "#         break\n",
    "#     k = i+1\n",
    "#     while k < len(permutation):\n",
    "#         current_subseq = []\n",
    "#         current_subseq.append(permutation[i])\n",
    "#         for j in range(k, len(permutation)):\n",
    "#             if current_subseq[-1] < permutation[j]:\n",
    "#                 current_subseq.append(permutation[j])\n",
    "#         incr_subseq.append(current_subseq)\n",
    "#         k += 1\n",
    "    \n",
    "# max_incr = sorted(incr_subseq, key=len)[-1]\n",
    "\n",
    "\n",
    "# print('find the longest decreasing subsequence')\n",
    "# # find the longest decreasing subsequence\n",
    "# decr_subseq = []\n",
    "# for i in range(len(permutation)):\n",
    "    \n",
    "#     if i+1 == (len(permutation)):\n",
    "#         decr_subseq.append(current_subseq)\n",
    "#         break\n",
    "#     k = i+1\n",
    "#     while k < len(permutation):\n",
    "#         current_subseq = []\n",
    "#         current_subseq.append(permutation[i])\n",
    "#         for j in range(k, len(permutation)):\n",
    "#             if current_subseq[-1] > permutation[j]:\n",
    "#                 current_subseq.append(permutation[j])\n",
    "#         decr_subseq.append(current_subseq)\n",
    "#         k += 1\n",
    "    \n",
    "# max_decr = sorted(decr_subseq, key=len)[-1]\n",
    "\n",
    "# with open('ouf.txt', 'w') as ouf:\n",
    "#     ouf.write(' '.join(max_incr))\n",
    "#     ouf.write('\\n')\n",
    "#     ouf.write(' '.join(max_decr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '>Rosalind_8407\\n'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-59-dcbaadf97b1c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     18\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[1;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'inf.txt'\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0minf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 20\u001b[1;33m     \u001b[0mn\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     21\u001b[0m     \u001b[0mpermutation\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     22\u001b[0m     \u001b[1;32mwhile\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '>Rosalind_8407\\n'"
     ]
    }
   ],
   "source": [
    "def find_max_subsequence(permutation):\n",
    "    n = len(permutation)\n",
    "    max_length = [1]*n            \n",
    "    for i in range(n):\n",
    "        for j in range(i):\n",
    "            if permutation[j] < permutation[i]:\n",
    "                max_length[i] = max(max_length[i], 1+max_length[j])\n",
    "    subsequence_length = max(max_length)\n",
    "    subsequence = []\n",
    "    for l in reversed(range(1,subsequence_length+1)):\n",
    "        for j in range(len(max_length)):\n",
    "            if l == max_length[j] and (len(subsequence) == 0 or permutation[j] < subsequence[-1]):\n",
    "                subsequence.append(permutation[j])\n",
    "                break\n",
    "    subsequence = subsequence[::-1]\n",
    "    return subsequence\n",
    "\n",
    "\n",
    "with open('inf.txt') as inf:\n",
    "    n = int(inf.readline())\n",
    "    permutation = []\n",
    "    while True:\n",
    "        my_list = list(map(int, inf.readline().strip().split()))\n",
    "        permutation.extend(my_list)\n",
    "        if len(my_list) == 0:\n",
    "            break\n",
    "\n",
    "max_subsequence = find_max_subsequence(permutation)\n",
    "            \n",
    "ans = find_max_subsequence(permutation[::-1])\n",
    "ans = ans[::-1]\n",
    "min_subsequence = ans\n",
    "\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    ouf.write(' '.join(list(map(str,max_subsequence))))\n",
    "    ouf.write('\\n')\n",
    "    ouf.write(' '.join(list(map(str,min_subsequence))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genome Assembly as Shortest Superstring\n",
    "\n",
    "Given: At most 50 DNA strings of approximately equal length, not exceeding 1 kbp, in FASTA format (which represent reads deriving from the same strand of a single linear chromosome).\n",
    "\n",
    "The dataset is guaranteed to satisfy the following condition: there exists a unique way to reconstruct the entire chromosome from these reads by gluing together pairs of reads that overlap by more than half their length.\n",
    "\n",
    "Return: A shortest superstring containing all the given strings (thus corresponding to a reconstructed chromosome)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make function to find suffix length\n",
    "def find_suffix_length(a,b):\n",
    "    suf_len = 0\n",
    "    min_len = min(len(a), len(b))\n",
    "    for i in range(min_len//2, min_len+1):\n",
    "        if a[-i:] == b[:i]:\n",
    "            suf_len = i\n",
    "            break\n",
    "    return suf_len\n",
    "\n",
    "\n",
    "# make a reads list from file\n",
    "name_DNA = {}\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            name = line[1:]\n",
    "            name_DNA[name] = ''\n",
    "        else:\n",
    "            name_DNA[name] += line\n",
    "\n",
    "reads = list(name_DNA.values())\n",
    "l = min(map(len, reads))//2\n",
    "n = len(reads)\n",
    "\n",
    "\n",
    "# make matrix of suffix length for reads\n",
    "len_matrix = [[0 for x in range(n)] for y in range(n)]\n",
    "for i, el_i in enumerate(reads):\n",
    "    for j, el_j in enumerate(reads):\n",
    "        if i == j:\n",
    "            len_matrix[i][j] = -1\n",
    "            continue\n",
    "        len_matrix[i][j] = find_suffix_length(el_i,el_j)\n",
    "\n",
    "\n",
    "# find possible pairs\n",
    "candidate_cells = [[(i,j) for j in range(n) if len_matrix[i][j] >= l] for i in range(n)]\n",
    "\n",
    "# find last string\n",
    "for i in range(len(candidate_cells)):\n",
    "    if len(candidate_cells[i]) == 0:\n",
    "        last_ind = i\n",
    "              \n",
    "# make list of row numbers in right order\n",
    "sequence = [last_ind]\n",
    "while len(sequence) < len(reads):\n",
    "    for i in range(len(candidate_cells)):\n",
    "        if i != last_ind and candidate_cells[i][0][1] == sequence[-1]:\n",
    "            sequence.append(candidate_cells[i][0][0])\n",
    "            break        \n",
    "sequence = sequence[::-1]\n",
    "row_numbers = sequence\n",
    "\n",
    "\n",
    "# combine strings to superstring\n",
    "shortest_superstring = ''\n",
    "for i in range(0, len(row_numbers)-1):\n",
    "    number = row_numbers[i]\n",
    "    next_number = row_numbers[i+1]\n",
    "    end_of_string = len_matrix[number][next_number]\n",
    "    shortest_superstring += reads[number][:-end_of_string]\n",
    "shortest_superstring += reads[row_numbers[-1]]\n",
    "\n",
    "\n",
    "# write answer to file\n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    ouf.write(shortest_superstring)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some challenge: find a mistake in this function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def find_max_overlap(len_matrix, candidate_cells, sequence, overlap):\n",
    "#     n = len(candidate_cells)\n",
    "#     if len(sequence) == n:\n",
    "#         return overlap, sequence\n",
    "    \n",
    "#     next_indices = []\n",
    "#     for ind_pair in candidate_cells[sequence[-1]]:\n",
    "#         next_indices.append(ind_pair[1])\n",
    "    \n",
    "#     max_overlap = overlap\n",
    "#     max_sequence = sequence\n",
    "#     for next_ind in next_indices:\n",
    "#         if next_ind not in sequence:\n",
    "#             new_overlap, new_sequence = find_max_overlap(len_matrix, candidate_cells, sequence + [next_ind], overlap + len_matrix[sequence[-1]][next_ind])\n",
    "#             if max_overlap < new_overlap:\n",
    "#                 max_overlap = new_overlap\n",
    "#                 max_sequence = new_sequence\n",
    "#     return max_overlap, max_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perfect Matchings and RNA Secondary Structures\n",
    "\n",
    "Given: An RNA string s of length at most 80 bp having the same number of occurrences of 'A' as 'U' and the same number of occurrences of 'C' as 'G'.\n",
    "\n",
    "Return: The total possible number of perfect matchings of basepair edges in the bonding graph of s."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "295950609069496384270872084480000000\n"
     ]
    }
   ],
   "source": [
    "# read RNA from file\n",
    "RNA = ''\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            continue\n",
    "        else:\n",
    "            RNA += line\n",
    "            \n",
    "#print(RNA)\n",
    "\n",
    "# recursion relation\n",
    "def total_number_perfect_matchings(n):\n",
    "    if n == 0:\n",
    "        return 1\n",
    "    else:\n",
    "        return (n)*total_number_perfect_matchings(n-1)\n",
    "\n",
    "\n",
    "# count perfect edges\n",
    "n1 = RNA.count('A')\n",
    "basepair_edges_AU = total_number_perfect_matchings(n1)\n",
    "n2 = RNA.count('C')\n",
    "basepair_edges_CG = total_number_perfect_matchings(n2)\n",
    "perfect_basepair_edges = basepair_edges_AU * basepair_edges_CG\n",
    "\n",
    "print(perfect_basepair_edges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Partial Permutations\n",
    "\n",
    "Given: Positive integers n and k such that 100≥n>0 and 10≥k>0.\n",
    "\n",
    "Return: The total number of partial permutations P(n,k), modulo 1,000,000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "85 8\n",
      "438400\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "n, k = [int(i) for i in input().split()]\n",
    "\n",
    "print(int(math.factorial(n)/math.factorial(n-k) % 1000000))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Introduction to Random Strings\n",
    "\n",
    "Given: A DNA string s of length at most 100 bp and an array A containing at most 20 numbers between 0 and 1.\n",
    "\n",
    "Return: An array B having the same length as A in which B[k] represents the common logarithm of the probability that a random string constructed with the GC-content found in A[k] will match s exactly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGAGTCTCGTAGAATATCGTATCGTTTAGCAGACCGGAATCTCGTTGCGACAACCTTTTCGCAAATTCCACTTGGCCCATGAGAGC\n",
      "0.099 0.111 0.158 0.237 0.279 0.338 0.394 0.422 0.475 0.520 0.593 0.629 0.662 0.711 0.784 0.812 0.859 0.904\n",
      "-71.369 -69.489 -63.934 -58.244 -56.279 -54.328 -53.154 -52.776 -52.405 -52.427 -53.126 -53.795 -54.620 -56.280 -60.018 -62.016 -66.462 -72.854\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "def find_log_prob(dna, gc_content):\n",
    "    symb_prob = {}\n",
    "    symb_prob['C'] = gc_content/2\n",
    "    symb_prob['G'] = gc_content/2\n",
    "    symb_prob['A'] = (1 - gc_content)/2\n",
    "    symb_prob['T'] = (1 - gc_content)/2\n",
    "    log_prob = 0\n",
    "    for symb in dna:\n",
    "        log_prob += math.log10(symb_prob[symb])\n",
    "    return log_prob\n",
    "\n",
    "\n",
    "s = input()\n",
    "A = list(map(float, input().strip().split()))\n",
    "\n",
    "B = [0 for i in range(len(A))]\n",
    "\n",
    "for i in range(len(A)):\n",
    "    B[i] = (find_log_prob(s, A[i]))\n",
    "    \n",
    "print(*[f'{x:0.3f}' for x in B])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enumerating Oriented Gene Orderings\n",
    "\n",
    "Given: A positive integer n≤6.\n",
    "\n",
    "Return: The total number of signed permutations of length n, followed by a list of all such permutations (you may list the signed permutations in any order)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "n = int(input())\n",
    "pairs = []\n",
    "\n",
    "import itertools\n",
    "\n",
    "numbers = list(range(1, n+1))\n",
    "permutations = list(itertools.permutations(numbers))\n",
    "signs_comb = list(itertools.product([1,-1],repeat = n))\n",
    "\n",
    " \n",
    "for item_1 in permutations:\n",
    "    for item_2 in signs_comb:\n",
    "        s = ''\n",
    "        for i in range(n):\n",
    "            s += str(item_1[i] * item_2[i]) + ' '\n",
    "        pairs.append(s)\n",
    "        \n",
    "with open('ouf.txt', 'w') as ouf:\n",
    "    ouf.write(str(len(pairs))+'\\n')\n",
    "    ouf.write('\\n'.join(pairs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finding a Spliced Motif\n",
    "\n",
    "Given: Two DNA strings s and t (each of length at most 1 kbp) in FASTA format.\n",
    "\n",
    "Return: One collection of indices of s in which the symbols of t appear as a subsequence of s. If multiple solutions exist, you may return any one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 17 22 24 26 42 46 51 53 56 67 70 74 76 87 90 92 94 96 98 105 111 114 117 121 124 126 128 133 138 141 143 148 150 152 158 162 165 169 177 180 193 200 210 214 216 225 228 233 236 238 241 244\n"
     ]
    }
   ],
   "source": [
    "dna = ''\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            s = dna\n",
    "            dna = ''\n",
    "        else:\n",
    "            dna += line\n",
    "t = dna\n",
    "            \n",
    "indices = []\n",
    "i = 0\n",
    "counter = 0\n",
    "while counter < len(t):\n",
    "    indices.append(s[i:].find(t[counter]) + i + 1)\n",
    "    i = indices[-1] + 1\n",
    "    counter += 1\n",
    "    \n",
    "print(*indices)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Transitions and Transversions\n",
    "\n",
    "Given: Two DNA strings s1 and s2 of equal length (at most 1 kbp).\n",
    "\n",
    "Return: The transition/transversion ratio R(s1,s2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.9021739130434783\n"
     ]
    }
   ],
   "source": [
    "name_DNA = {}\n",
    "with open('inf.txt') as inf:\n",
    "    for line in inf:\n",
    "        line = line.strip()\n",
    "        if line[0] == '>':\n",
    "            name = line[1:]\n",
    "            name_DNA[name] = ''\n",
    "        else:\n",
    "            name_DNA[name] += line\n",
    "\n",
    "dna_list = list(name_DNA.values())\n",
    "\n",
    "transition = {'A':'G', 'G':'A', 'C':'T', 'T':'C'}\n",
    "\n",
    "\n",
    "def transition_transversion_ratio(s1, s2):\n",
    "    n = len(s1)\n",
    "    transi = 0\n",
    "    transv = 0\n",
    "    for i in range(n):\n",
    "        if s1[i] == s2[i]:\n",
    "            continue\n",
    "        elif transition[s1[i]] ==  s2[i]:\n",
    "            transi += 1\n",
    "        else:\n",
    "            transv += 1\n",
    "    R = transi/transv\n",
    "    return R\n",
    "\n",
    "print(transition_transversion_ratio(dna_list[0], dna_list[1]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Completing a Tree\n",
    "\n",
    "Given: A positive integer n (n≤1000) and an adjacency list corresponding to a graph on n nodes that contains no cycles.\n",
    "\n",
    "Return: The minimum number of edges that can be added to the graph to produce a tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('inf.txt') as inf:\n",
    "    n = int(inf.readline())\n",
    "    nodes = {}\n",
    "    for i in range(1, n+1):\n",
    "        nodes[i] = []\n",
    "    while True:\n",
    "        line = inf.readline().strip().split()\n",
    "        if not line:\n",
    "            break\n",
    "        x, y = (map(int, line))\n",
    "        nodes[x].append(y)\n",
    "        nodes[y].append(x)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_cluster(nodes, current_node, cluster=set()):\n",
    "#     print(cluster)\n",
    "    cluster.add(current_node)\n",
    "    children = [c for c in nodes[current_node] if c not in cluster]\n",
    "    if len(children) == 0:\n",
    "        return cluster\n",
    "    else:\n",
    "        for c in children:\n",
    "            cluster.update(find_cluster(nodes, c, cluster))\n",
    "        return cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43\n"
     ]
    }
   ],
   "source": [
    "used_nodes = set()\n",
    "n_clusters = 0\n",
    "for node in nodes.keys():\n",
    "    if node not in used_nodes:\n",
    "        cluster = find_cluster(nodes, node, cluster=set())\n",
    "        used_nodes.update(cluster)\n",
    "        n_clusters += 1\n",
    "        \n",
    "print(n_clusters - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[3, 2, 4], [5, 4], [1]]\n",
      "[3, 2, 4, 5, 4, 1]\n"
     ]
    }
   ],
   "source": [
    "# l = max(map(len,(nodes.values())))\n",
    "# groups = []\n",
    "# used_nodes = []\n",
    "# all_count = False\n",
    "# while l >= 0 and not all_count:\n",
    "#     for key, value in nodes.items():\n",
    "#         if len(value) == l and key not in used_nodes:\n",
    "#             new_list = [key]\n",
    "#             new_list.extend(value)\n",
    "#             groups.append(new_list)\n",
    "#             used_nodes.extend(new_list)\n",
    "#     l -= 1   \n",
    "# print(groups)\n",
    "# print(used_nodes)\n",
    "# print(len(groups)-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
